### PR TITLE
add main processing steps

### PR DESCRIPTION
**Pull request by oberonia78**
This PR addresses 'initial_threshold', 'fuzzy_logic_computation', 'bimodality_test', and 'refine_with_ancillary' tests. They are all unique and newly developed, so they may need careful review. 

For an optimal review process, I recommend the following order:

1) 'refine_with_bimodality'
2) 'fuzzy_value_computation'
3) 'initial_threshold'
4) 'masking with ancillary'."

### CODE REVIEW

File: src/dswx_sar/defaults/algorithm_parameter_s1.yaml [hunk 1]
```diff
@@ -50,6 +50,34 @@ runconfig:
...
+            # membership bound for reference water                
+            reference_water:
+                # Minimum reference water value for membership
+                member_min: 0.8
+                # Maximum reference water value for membership
+                member_max: 0.95
+            # Dark area is defined where cross-pol is lower than cross_land
+            # Water is defined where cross-pol is lower than cross_water
+            dark_area:
+                # Threshold for land in the dark area definition
```
Comment (seongsujeong) [inline src/dswx_sar/defaults/algorithm_parameter_s1.yaml:69]: Please indicate the unit for the threshold (dB)

File: src/dswx_sar/initial_threshold.py [hunk 2]
```diff
@@ -0,0 +1,1826 @@
...
+    logger.info(f'Start Initial Threshold')
+
+    processing_cfg = cfg.groups.processing
+    pol_list = processing_cfg.polarizations
+    pol_all_str = '_'.join(pol_list)
+
+    outputdir = cfg.groups.product_path_group.scratch_path
+    init_threshold_cfg = processing_cfg.initial_threshold
+    number_iterations = init_threshold_cfg.number_iterations
+    number_workers = init_threshold_cfg.number_cpu
```
Comment (oberonia78) [inline src/dswx_sar/initial_threshold.py:1487]: need to check OMP_number.

File: src/dswx_sar/initial_threshold.py [hunk 3]
```diff
@@ -0,0 +1,1826 @@
...
+    wbd_gdal = gdal.Open(wbd_im_str)
+
+    # compute the statistics of intensity.
+    ref_water_max = processing_cfg.reference_water.max_value
+    wbd_whole = wbd_gdal.ReadAsArray()
+    wbd_whole_norm = wbd_whole / ref_water_max * 100
+
+    # create water masks for normal,
+    # flood and drought using dilation and erosion
+    water_mask_set = create_three_watermask(wbd_whole,
```
Comment (oberonia78) [inline src/dswx_sar/initial_threshold.py:1515]: expose the variable for the spatial shift for flood and drought.

File: src/dswx_sar/initial_threshold.py [hunk 4]
```diff
@@ -0,0 +1,1826 @@
...
+                # All thresholds can be accepted.
+                thres_max[band_ind] = 30
+            else:
+                # extract intensity values from the water areas
+                int_water_array = intensity_whole[
+                    band_ind, (wbd_whole_norm > 90) &
+                    (np.nansum(initial_water_set, axis=2) > 1)]
+
+                int_water_array = remove_invalid(int_water_array)
+
```
Comment (oberonia78) [inline src/dswx_sar/initial_threshold.py:1552]: add more explanation

File: src/dswx_sar/initial_threshold.py [hunk 5]
```diff
@@ -0,0 +1,1826 @@
...
+        if win_size < minimum_pixel_number:
+            logger.info('tile & image sizes are too small')
+            coordinate = []
+            return coordinate
+
+        # Check if number of pixel is enough
+        num_pixel_max = win_size * win_size / 3
+
+        number_y_window = np.int16(height / win_size)
```
Comment (oberonia78) [inline src/dswx_sar/initial_threshold.py:296]:
```suggestion
        number_y_window = np.ceil(height / win_size)
```

File: src/dswx_sar/initial_threshold.py [hunk 6]
```diff
@@ -0,0 +1,1826 @@
...
+
+    if dswx_workflow.lower() == 'twele':
+        tile_selection_method = 'twele'
+
+    # Tile Selection (w/o water body)
+    if (height < winsize) | (width < winsize):
+        logger.info('winsize is smaller than image size')
+        winsize = np.min([winsize, height, width])
+
+    number_y_window = np.int16(height / winsize)
```
Comment (oberonia78) [inline src/dswx_sar/initial_threshold.py:1330]:
```suggestion
    number_y_window = np.ceil(height / winsize)
```

File: src/dswx_sar/initial_threshold.py [hunk 7]
```diff
@@ -0,0 +1,1826 @@
...
+
+    threshold_tau_set = []
+    mode_tau_set = []
+    candidate_coord_set = []
+
+    tile_selection_object = TileSelection(ref_water_max=water_cfg.max_value,
+                                          no_data=water_cfg.no_data_value)
+    ## Tile Selection (with water body)
+    for polind, pol in enumerate(pol_list):
+        if pol == 'ratio':
```
Comment (oberonia78) [inline src/dswx_sar/initial_threshold.py:1346]:
```suggestion
```

File: src/dswx_sar/initial_threshold.py [hunk 8]
```diff
@@ -0,0 +1,1826 @@
...
+            # TBD
+            thresh_list = [0.5, 0.7, 0.96]
+        else:
+            thresh_list = [0.09, 0.8, 0.97]
+
+        tile_selection_object.threshold_twele = thresh_list
+
+        candidate_coord = tile_selection_object.tile_selection_wbd(
+                           intensity=intensity[polind, :, :],
+                           water_mask=wbdsub,
```
Comment (oberonia78) [inline src/dswx_sar/initial_threshold.py:1356]:
```suggestion
                           water_mask=ref_water_body_sub,
```

File: src/dswx_sar/initial_threshold.py [hunk 9]
```diff
@@ -0,0 +1,1826 @@
...
+            target_im = np.squeeze(10*np.log10(intensity[polind, :, :]))
+        else:
+            min_im, max_im, step_im = -1000, -1000, 0.1
+            target_im = np.squeeze(intensity[polind, :, :])
+
+        if len(candidate_coord) > 0 :
+
+            wbdsub_norm = wbdsub/tile_selection_object.wbd_max_value
+            if wbdsub_norm.shape[0]*wbdsub_norm.shape[1] == 0:
+                water_variation = 0
```
Comment (oberonia78) [inline src/dswx_sar/initial_threshold.py:1375]: add comments

**Review by seongsujeong — COMMENTED** [review]
Here are feedbacks for `fuzzy_value_computation.py` I have not tested this PR yet because I do not have runconfig for this - Seongsu

File: src/dswx_sar/fuzzy_value_computation.py [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+
+    Parameters
+    ----------
+    intensity : numpy.ndarray
+        The intensity or ratio layers. A number of layers could be arbitrary.
+    slope : numpy.ndarray
+        The slope angle raster
+    hand : numpy.ndarray
+        The Height Above Nearest Drainage (HAND) raster
+    hand_threshold : float
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/fuzzy_value_computation.py:165]: Great. Thank you for pointing out.

File: src/dswx_sar/fuzzy_value_computation.py [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+    '''
+    center_value = (minv + maxv) / 2
+    output = np.zeros(np.shape(values))
+    
+    membership_left = 1 - 2 * (((values - minv) / (maxv - minv))**2)
+    mask_left = (values >= minv) & (values <= center_value)
+    output[mask_left] = membership_left[mask_left]
+
+    membership_right = 2 * (((values - maxv) / (maxv - minv))**2)
+    mask_right = (values >= minv) & (values <= center_value)
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/fuzzy_value_computation.py]: Thank you for finding out about this issue. I'm now investigating when this happened. But I'm suspecting that the code was corrupted while I transfer the code from mine to github.

File: src/dswx_sar/dswx_s1.py [hunk 12]
```diff
@@ -22,12 +32,30 @@ def dswx_s1_workflow(cfg):
     # Create mosaic burst RTCs
     mosaic_rtc_burst.run(cfg)
 
-    # preprocessing (relocating ancillary data and filtering)
-    pre_processing.run(cfg)
```
Comment (seongsujeong) [inline src/dswx_sar/dswx_s1.py:26]: Any reason that we're not doing the preprocessing? `initial_threshold.run(cfg)` attempts to load filtered image [here](https://github.com/oberonia78/DSWX-SAR/blob/10134e5377c8e5415e079e61fae90e8a995f5572/src/dswx_sar/initial_threshold.py#L1496), and I think we need preprocessing?

Comment (oberonia78) [reply src/dswx_sar/dswx_s1.py:26]: Yes. You are right. It should be included.

File: src/dswx_sar/initial_threshold.py [hunk 13]
```diff
@@ -0,0 +1,1827 @@
...
+        ----------
+        raster : numpy.ndarray
+        min_value : float
+        max_value : float
+
+        Returns
+        -------
+        numpy.ndarray
+        """
+        return (raster - min_value) / (max_value - min_value)
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/initial_threshold.py:67]: Different users might employ a variety of reference water maps. These can range from water occurrence maps with values from 0 to 100, to seasonality maps with a scale of 0-12, or even binary water maps with values of either 0 or 1. By setting a maximum value, we can use all these options seamlessly. Once this is done, we can identify permanent water by applying a threshold to the reference water. Without this rescaling step, we'd need to use different thresholds for seasonality and occurrence, which is not our intended approach.

[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/initial_threshold.py:67]: ah. I was confused. I thought this was related to the water map. In Twele's paper, it was mentioned that the intensity was rescaled to 0 -1. I followed the instruction, but you may be right.

File: src/dswx_sar/initial_threshold.py [hunk 14]
```diff
@@ -0,0 +1,1889 @@
...
+        # distances, indices = tree.query(grid_points, k=number_point)  # find the 3 nearest neighbors
+        # # Inverse distance weighting
+        # weights = 1.0 / distances
+        # weights /= np.sum(weights, axis=1, keepdims=True)
+        # # weights /= weights.sum(axis=1)[:, np.newaxis]  # normalize weights
+
+        # grid_values = np.sum(values[indices] * weights, axis=1)
+
+        # # Now we can reshape the result back to the original grid shape
+        # grid_values = grid_values.reshape(grid_x.shape)
```
Comment (LiangJYu) [inline src/dswx_sar/initial_threshold.py:1265]:
```suggestion
```
Clean up commented code

File: src/dswx_sar/initial_threshold.py [hunk 15]
```diff
@@ -0,0 +1,1889 @@
...
+                    metric_obj = refine_with_bimodality.BimodalityMetrics(
+                        int_water_array)
+                    bimodal_bool = metric_obj.compute_metric()
+                    if bimodal_bool:
+                        logger.info('Bright water found on images')
+                        logger.info(f'      :{metric_obj.first_mode[0]:.2f} '
+                                    f'vs. {metric_obj.second_mode[0]:.2f}')
+
+                        int_sub_mean = metric_obj.first_mode[0]
+                        int_sub_std = metric_obj.first_mode[1]
```
Comment (LiangJYu) [inline src/dswx_sar/initial_threshold.py:1629]:
```suggestion
                        int_sub_mean, int_sub_std = metric_obj.first_mode
```

File: src/dswx_sar/initial_threshold.py [hunk 16]
```diff
@@ -0,0 +1,1889 @@
...
+                logger.info(f'thres intensity [dB] over water {pol_list[band_ind]}:'
+                            f' {thres_max[band_ind]:.2f}, {bimodal_bool}')
+
+        block_row = init_threshold_cfg.maximum_tile_size.y
+        block_col = init_threshold_cfg.maximum_tile_size.x
+
+        # number_y_window
+        n_rows_block = np.int16(height / block_row)
+        # number_x_window
+        n_cols_block = np.int16(width / block_col)
```
Comment (LiangJYu) [inline src/dswx_sar/initial_threshold.py:1665]: Like above, `ceil` these? Or `//`?

File: src/dswx_sar/initial_threshold.py [hunk 17]
```diff
@@ -0,0 +1,1889 @@
...
+        #             figname=f'ki_tau_{iter_ind}iter_')
+        #     else:
+        #         for band_ind2 in range(band_number):
+        #             vs.block_threshold_visulaization(
+        #                 np.squeeze(intensity[band_ind2, :, :]),
+        #                 block_row,
+        #                 block_col,
+        #                 threshold_tau_set[:, :, band_ind2],
+        #                 outputdir,
+        #                 f'ki_tau_{pol_list[band_ind2]}_{iter_ind}')
```
Comment (LiangJYu) [inline src/dswx_sar/initial_threshold.py:1854]:
```suggestion
```
Remove commented code

**Review by LiangJYu — APPROVED** [review]
LGTM. Nothing deal breaking. Perhaps at a later time breaking apart into smaller chunks may be a good idea.

File: src/dswx_sar/fuzzy_value_computation.py [hunk 18]
```diff
@@ -0,0 +1,505 @@
...
+    mu_h = np.nanmean(hand[initial_map == 1])
+    std_h = np.nanstd(hand[initial_map == 1])
+    min_dem = mu_h
+    max_dem = mu_h + (std_h + 2.5) * std_h
+    hand[np.isnan(hand)] = 0
+
+    if np.isnan(min_dem):
+        min_dem = 0
+    if np.isnan(max_dem):
+        max_dem = 15
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/fuzzy_value_computation.py:327]: I will expose the value in the runconfig. But the min and max values are automatically computed from the HAND, but in the case that `max_dem` and `min_dem` are `nan`, then an alternative value is used.

File: src/dswx_sar/initial_threshold.py [hunk 19]
```diff
@@ -0,0 +1,1889 @@
...
+                               threshold=0.7,
+                               min_im=-30,
+                               max_im=5,
+                               numstep=100):
+        ''' Select tiles with bimodal distribution
+
+        Parameters
+        ----------
+        intensity : numpy.ndarray
+            intensity array [dB]
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/initial_threshold.py:83]: You are correct. Thank you 👍

File: src/dswx_sar/fuzzy_value_computation.py [hunk 20]
```diff
@@ -0,0 +1,505 @@
...
+    sobely = cv2.Sobel(dem, cv2.CV_64F, 0, 1, ksize=SOBEL_KERNEL_SIZE)  # y
+
+    # Compute slope
+    slope_angle = np.arctan(np.sqrt(
+        (sobelx / SOBEL_KERNEL_SIZE / PIXEL_RESOLUTION_X) ** 2 +
+        (sobely / SOBEL_KERNEL_SIZE / PIXEL_RESOLUTION_Y) ** 2)) * RAD_TO_DEG
+    return slope_angle
+
+
+def smf(values, minv, maxv):
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/fuzzy_value_computation.py:46]: Yes. It is exactly same.

File: src/dswx_sar/initial_threshold.py [hunk 21]
```diff
@@ -0,0 +1,1889 @@
...
+        # create 3 dimensional water mask
+        if water_number == 2:
+            water_mask = np.expand_dims(water_mask, axis=0)
+        _, water_nrow, water_ncol = water_mask.shape
+
+        # If image size is smaller than window size,
+        # window size become half of image size
+        if (height <= win_size) | (width <= win_size):
+            win_size0 = win_size
+            win_size = int(win_size / 2)
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/initial_threshold.py:300]: Good point. Thank you.

File: src/dswx_sar/initial_threshold.py [hunk 22]
```diff
@@ -0,0 +1,1889 @@
...
+
+class TileSelection:
+    '''
+    Select tile candidates that have boundary between water and non-water
+    '''
+    def __init__(self, ref_water_max, no_data):
+        # no data value for reference_water
+        self.no_data = no_data
+        # value for pixels where water exists in reference_water
+        self.wbd_max_value = ref_water_max
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/initial_threshold.py:46]: `wbd` represents `water body database`

File: src/dswx_sar/initial_threshold.py [hunk 23]
```diff
@@ -0,0 +1,1889 @@
...
+def compute_ki_threshold(intensity, min_im, max_im, step_im):
+    """ Computes the threshold using Kittler-Illingworth algorithm
+
+    Parameters
+    ----------
+    intensity: np.ndarray
+        The image intensities.
+    min_im: int
+        The minimum intensity.
+    max_im: int
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/initial_threshold.py:606]: I changed them to `min_intensity_histogram`, `max_intensity_histogram`

File: src/dswx_sar/initial_threshold.py [hunk 24]
```diff
@@ -0,0 +1,1889 @@
...
+    counts, bins = np.histogram(intensity,
+                                bins=np.linspace(min_im,
+                                                 max_im,
+                                                 numstep+1),
+                                density=True)
+    bins2 = bins[:-1]
+
+    prob_array = np.zeros_like(bins2)
+
+    for bin_ind, tau in enumerate(bins2):
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/initial_threshold.py:632]: Great. Thanks for letting me know. I'll change the code as you suggested.

File: src/dswx_sar/refine_with_bimodality.py [hunk 25]
```diff
@@ -0,0 +1,1181 @@
...
+                                       ashman_threshold],
+                            meta_info=im_meta,
+                            outputdir=outputdir,
+                            input_dict=input_file_dict)
+
+    water_bindary[fill_gap_bindary] = True
+
+    water_meta = dswx_sar_util.get_meta_from_tif(filt_im_str)
+
+    mask_excluded = water_mask_image == 4
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/refine_with_bimodality.py:1120]: Good point. I replaced the line with below.
```suggestion
    landcover_mask_excluded = \
        water_mask_image == band_assign_value_dict['landcover_mask']```

File: src/dswx_sar/refine_with_bimodality.py [hunk 26]
```diff
@@ -0,0 +1,1181 @@
...
+    dark_land = np.logical_and(mask_excluded, water_bindary)
+    no_data_raster = np.isnan(
+        np.squeeze(np.nanmean(band_set, axis=0)))
+    no_data_raster = no_data_raster | (water_mask_image == 255)
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/refine_with_bimodality.py:1124]:
```suggestion
    no_data_raster = no_data_raster | \
        (water_mask_image == band_assign_value_dict['no_data'])```

I changed the line as above.

File: src/dswx_sar/fuzzy_value_computation.py [hunk 27]
```diff
@@ -0,0 +1,505 @@
...
+    slope_z = zmf(slope,
+                  fuzzy_option['slope_min'],
+                  fuzzy_option['slope_max'])
+
+    # compute area membership
+    logger.info('area s membership')
+    handem = hand < fuzzy_option['hand_threshold']
+    wbsmask = (initial_map == 1) & (handem)
+    watermap = calculate_water_area(wbsmask)
+    area_s = smf(watermap, 5, 40)
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/fuzzy_value_computation.py:340]: They are min and max values for the initial water area. I'll expose them in runconfig.

File: src/dswx_sar/fuzzy_value_computation.py [hunk 28]
```diff
@@ -0,0 +1,505 @@
...
+    reference_water_s = smf(reference_water,
+                            fuzzy_option['reference_water_min'],
+                            fuzzy_option['reference_water_max'])
+
+    # compute fuzzy-logic-based value
+    method_dict = {
+        'opera_dswx_s1': lambda: (nansum_intensity_z_set / (band_number) * 0.5 + \
+                   (hand_z + slope_z + reference_water_s)  / 3 * 0.5),
+        'twele': lambda: (nansum_intensity_z_set / (band_number + 1) / 2 + \
+                   (hand_z + slope_z)  / 3 / 2)
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/fuzzy_value_computation.py]: I added the following comments.
    # The Opera dswx s1 algorithm calculates fuzzy-logic-based values based on
    # input parameters, including intensity, hand, slope, and reference water.
    # The Twele algorithm, on the other hand, computes these values using
    # intensity, hand, slope, and areas.
    # It's important to note that half of the fuzzy values (0.5) are derived from
    # the intensity values, while the remaining half (0.5) comes from ancillary data.
    # To achieve this, the intensity membership is divided by the number of bands
    # and then multiplied by 0.5. The maximum value for intensity membership is
    # capped at 0.5. Similarly, the membership of the ancillary data contributes
    # 0.5 to the final result.`

File: src/dswx_sar/schemas/algorithm_parameter_s1.yaml [hunk 29]
```diff
@@ -50,19 +57,59 @@ runconfig:
...
             seed: num(min=0, max=1, required=False)
             # value where region growing is stopped
             tolerance: num(min=0, max=1, required=False)
             line_per_block: num(min=1, required=False)
-    
+
+        masking_ancillary:
+            vv_threshold: num(min=-30, max=10, required=False)
+            vh_threshold: num(min=-30, max=10, required=False)
+            water_threshold: num(min=0, max=100, required=False)
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/schemas/algorithm_parameter_s1.yaml]: I added the explanation of the above parameters.

**Pull request merged**